### PR TITLE
Language tree

### PR DESCRIPTION
**Pull request by steelsojka**
This PR is the revised work of this one https://github.com/neovim/neovim/pull/13207.

This PR implements the language tree for language injection in treesitter.

97899023-56e43780-1cfe-11eb-92b6-0707ac1abd0c

The above example includes 5 different parsers and highlight injections.

- JSDoc highlighting
- Whole file javascript highlighting
- HTML template highlighting
- The class attribute is injected with another javascript parser
- CSS style highlighting in the styles tag

The parser/language tree can parse separate areas of the file together or in isolation (jsdoc).
This will enable highlighting for embedded languages like ERB and EJS templates. For example,

```ejs
<% for (var i; i <10; i++) { %>
  <li><%= i %></li>
<% } %>
```

The syntax tree for the javascript parser would only see these regions

```js
for (var i; i < 10; i++) {
  i
}
```

In this case each region is parsed within the same parsing context.

For injections like JSDoc, we don't want to parse all jsdocs together, because this can lead to parsing errors in the resulting tree. We can now isolate different regions so they are parsed within different contexts but still only using a single parser.

Also, allowing the a custom for the parser allows us (nvim-treesitter in this case) to do recursive language injection javascript -> html -> javascript

### CODE REVIEW

**Comment by steelsojka** [comment]
I will fix the broken tests and add new ones for the new functionality.

**Comment by vigoux** [comment]
Yay ! Thanks for your awesome work ! 
While we worked together on this for some part, I'll review and test all of this as soon as possible !

**Comment by bfredl** [comment]
Cool stuff, I will also take a look and try to understand how it works :]

**Comment by steelsojka** [comment]
The other thing this needs is a predicate to offset the injection range. I need to enable the parser to accept ranges instead of nodes. Then we can compute the ranges we need to include for the injected language. This will let us strip off quotes from strings with injected languages and even strip out leading parts of a line... Imagine `@example` tag in a jsdoc getting highlighted with JS.

**Comment by vigoux** [comment]
Pushed some partial test fixes, still highlight tests to fix, that are not working for some reason.

**Comment by steelsojka** [comment]
I have all the current tests fixed. I need to write tests for the injections. @vigoux How should I go about testing injection if there is only the C parser included in nvim? I need to inject another language or C within C (which I could do easily). I was thinking of just injecting C int a C string if we can't include something like a lua parser...

**Comment by theHamsta** [comment]
@steelsojka you can inject C in C macros.

This is what I was wanting also for our C/C++ highlighting. The macro highlighting. Should in the ideal case have some background highlighting to distinguish macro C from normal C

**Comment by steelsojka** [comment]
@theHamsta That will work great! I really need https://github.com/neovim/neovim/pull/13270 to be merged in before it will full work. Right now the line breaks will be sent to the injected parser. The above will give me the flexibility to implement the predicates to strip any characters (like quotes from strings). I'll give a try first... maybe the parser will handle it in this case.

**Comment by steelsojka** [comment]
Screenshot_20201115_071155

Looks like this scenario works fine in C... Will need that PR for other languages though.

File: test/functional/lua/treesitter_spec.lua
Comment (theHamsta) [inline test/functional/lua/treesitter_spec.lua:829]: I always wondered whether all these globals in all the tests might infer with each other when there is a small typo and a variable from a different test is used.

Comment (steelsojka) [reply test/functional/lua/treesitter_spec.lua:829]: Yeah, I'm not sure in what context these get executed in. I would assume each test is it's fresh nvim instance but I may be wrong on that.

Comment (theHamsta) [reply test/functional/lua/treesitter_spec.lua:829]: I also always assumed that. But always when something weird was going on this made be feel uncomfortable.

**Review by vigoux — APPROVED** [review]
That looks fairly good to me, waiting for the C injection then !

File: test/functional/lua/treesitter_spec.lua
Comment (steelsojka) [reply test/functional/lua/treesitter_spec.lua:829]: I can test it and see if anything carries from test to test.

Comment (bfredl) [reply test/functional/lua/treesitter_spec.lua:829]: code that exists inside an `it()` block doesn't "carry over". Only vars defined in `before_each()` blocks is shared between multiple test instances.

Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua:829]: Those locals will stay local to this test, I am pretty sure

Comment (theHamsta) [reply test/functional/lua/treesitter_spec.lua:829]: I'm referring to the globals in the lua_exec block. But it's probably a Neovim instance per it block and thus a reset of the Lua VM.

Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua:829]: That's what I am talking about too, I think they are local per `it` call

**Comment by Shougo** [comment]
I want to use the feature in context_filetype.vim.

https://github.com/Shougo/context_filetype.vim

File: runtime/lua/vim/treesitter/highlighter.lua
Comment (vigoux) [inline runtime/lua/vim/treesitter/highlighter.lua:67]: That's not exactly right, can we check here that the `tree._source` is actually a `bufnr` ?

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (vigoux) [inline runtime/lua/vim/treesitter/languagetree.lua:104]: Here, this disabled incremental parsing, we need to have an association between trees and regions, to be able to edit the trees, and parse them incrementaly.

**Review by vigoux — CHANGES_REQUESTED** [review]

**Comment by tami5** [comment]
Looking forward for this PR good work

File: runtime/lua/vim/treesitter/languagetree.lua
Comment (vigoux) [reply runtime/lua/vim/treesitter/languagetree.lua]: Should be good now, but that caused test failures, resolving for now.

**Comment by vigoux** [comment]
Hey @bfredl @theHamsta, I think we are pretty much ready.

Injection is working, but we had to do a little API break because of the shift in what parser actually mean now.
`Parser` have been changed to `LanguageTree`, and a `LanguageTree` is guaranteed unique per buffer.

We had to simplify one thing : incremental parsing is disabled for injected languages _for now_ because of the difficulty of the task.

For root languages, incremental parsing is done as usual.

If you can, please review this.

And a huuuge thanks to @steelsojka to do this.

**Comment by bfredl** [comment]
> We had to simplify one thing : incremental parsing is disabled for injected languages for now because of the difficulty of the task.

I take it that it is still lazy to changes of a injected language range as a _whole_ ? I mean if you leave a markdown codeblock unchanged then it won't be reparsed, but any change inside that block might reparse the entire block.

**Comment by vigoux** [comment]
Nah we are reparsing everything from scratch.
Detecting that the injection did not changed is a headache.

**Comment by steelsojka** [comment]
I think the important thing right here is that it's still incrementally parsing for the root language which is what the existing functionality covers. SInce injected languages are currently not implemented, it's not breaking anything by skipping the incremental parsing on injected languages (for now). All the functionality still works, but a potential performance improvement can be made. I've been using this PR for development for a couple weeks now on a very large app and I haven't seen any large performance decreases.

**Comment by bfredl** [comment]
ok, if injected language are "not supported" yet that makes sense I guess :D

**Comment by steelsojka** [comment]
@bfredl I wouldn't consider them "not supported". They are fully functional. I would consider the incremental parsing portion of it a performance improvement. From a user perspective, it will operate the same way with or without it. If anything it's in "beta".

**Comment by bfredl** [comment]
well the exact wording you used were "since injected languages are currently not implemented" but maybe this was not intended.

**Comment by steelsojka** [comment]
Sorry, that's my fault. It was a poor choice of words. What I meant was since it's not currently implemented in neovim master it won't hinder any existing functionality.

**Comment by bfredl** [comment]
yes, that makes sense.

File: test/functional/lua/treesitter_spec.lua
Comment (bfredl) [inline test/functional/lua/treesitter_spec.lua:645]: typically we will share the screen initiation in a `before_each()` block (with a `describe()` to group the highlight related tests)

**Comment by bfredl** [comment]
In general this looks good to me. Not sure what is going on with the new stuff (I trust you people on that) but the changes to the "old" API:s looks OK to me :)

File: test/functional/lua/treesitter_spec.lua
Comment (steelsojka) [reply test/functional/lua/treesitter_spec.lua]: Done!

**Comment by theHamsta** [comment]
I have also not to much knowledge about this code base. So I only had a brief look over the code. It looks good!

**Comment by steelsojka** [comment]
@vigoux Can you restart the checks on this PR (I'm not sure if I can do it)? I think one of them stalled.

**Comment by vigoux** [comment]
I restarted the GHA, once that goes green I'll hit that merge button

**Comment by theHamsta** [comment]
Gitlab's "Merge when pipeline succeeds" would be good here.

**Review by vigoux — APPROVED** [review]

**Pull request merged**
